src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/OptionsBuilderExtensionsTests.cs

```diff
@@ -0,0 +1,359 @@
+            });
+
+            using (var host = hostBuilder.Build())
+            {
+                var error = await Assert.ThrowsAsync<OptionsValidationException>(async () =>
+                {
+                    await host.StartAsync();
+                });
+
+                ValidateFailure<ComplexOptions>(error, 1);
+            }
+        }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/34580", TestPlatforms.Windows, TargetFrameworkMonikers.Netcoreapp, TestRuntimes.Mono)]
+        public async Task ValidateOnStart_ErrorMessageSpecified_FailsWithCustomError()
+        {
+            var hostBuilder = CreateHostBuilder(services =>
+            {
+                services.AddOptions<ComplexOptions>()
+                    .Configure(o => o.Boolean = false)
+                    .Validate(o => o.Boolean, "Boolean must be true.")
+                    .ValidateOnStart();
+            });
+
+            using (var host = hostBuilder.Build())
+            {
+                var error = await Assert.ThrowsAsync<OptionsValidationException>(async () =>
+                {
+                    await host.StartAsync();
+                });
+
+                ValidateFailure<ComplexOptions>(error, 1, "Boolean must be true.");
+            }
+        }
+
+        internal class FakeService { }
+
+        internal class FakeSettings
+        {
+            public string Name { get; set; }
+        }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/34580", TestPlatforms.Windows, TargetFrameworkMonikers.Netcoreapp, TestRuntimes.Mono)]
+        public async Task ValidateOnStart_NamedOptions_ValidatesFailureOnStart()
+        {
+            var hostBuilder = CreateHostBuilder(services =>
+            {
+                services.AddOptions().AddSingleton(new FakeService());
+                services
+                    .AddOptions<FakeSettings>("named")
+                    .Configure<FakeService>((o, _) =>
+                    {
+                        o.Name = "named";
+                    })
+                    .Validate(o => o.Name == null, "trigger validation failure for named option!")
+                    .ValidateOnStart();
+            });
+
+            using (var host = hostBuilder.Build())
+            {
+                var error = await Assert.ThrowsAsync<OptionsValidationException>(async () =>
+                {
+                    await host.StartAsync();
+                });
+
+                ValidateFailure<FakeSettings>(error, 1, "trigger validation failure for named option!");
+            }
+        }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/34580", TestPlatforms.Windows, TargetFrameworkMonikers.Netcoreapp, TestRuntimes.Mono)]
+        private async Task ValidateOnStart_AddOptionsMultipleTimesForSameType_LastOneGetsTriggered()
+        {
+            bool firstOptionsBuilderIgnored = true;
```

Comment:
(nit) I would make both of these booleans use the same pattern. To me it is more understandable to have:

```C#
bool firstOptionsBuilderTriggered = false;
bool secondOptionsBuilderTriggered = false;


Assert.False(firstOptionsBuilderTriggered);
Assert.True(secondOptionsBuilderTriggered);
```